lib/galaxy/tools/actions/__init__.py

```diff
@@ -500,6 +502,21 @@ def get_output_name( self, output, dataset, tool, on_text, trans, incoming, hist
         else:
             return self._get_default_data_name( dataset, tool, on_text=on_text, trans=trans, incoming=incoming, history=history, params=params, job_params=job_params )
 
+    def set_metadata_defaults( self, output, dataset, tool, on_text, trans, incoming, history, params, job_params ):
+        """
+        This allows to map names of input files to metadata default values. Example:
+
+        <data format="tabular" name="output" label="Tabular output, aggregates data from individual_inputs" >
+            <actions>
+                <action name="column_names" type="metadata" default="${','.join([input.name for input in $individual_inputs ])}" />
+            </actions>
+        </data>
+        """
+        if output.actions:
+            for action in output.actions.actions:
+                if action.tag == "metadata":
+                    action.default = fill_template( action.default, context=params )
```

Comment:
My instinct is that you need to move all this out of ``galaxy/tool/actions/__init__.py`` (happens before the job) and into ``finish()`` of ``JobWrapper`` in ``jobs/__init__.py`` (happen after the job), but I'm not sure exactly.